src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -602,18 +595,79 @@ private bool IsPropertyASink(IPropertyReferenceOperation propertyReferenceOperat
 
             private IEnumerable<IArgumentOperation> GetTaintedArguments(ImmutableArray<IArgumentOperation> arguments)
             {
+                foreach (IArgumentOperation arg in arguments)
+                {
+                    UpdateAbstractValueForArrayBeforeEnteringSinkForArray(arg, arg.Value);
+                }
+
                 return arguments.Where(
                     a => this.GetCachedAbstractValue(a).Kind == TaintedDataAbstractValueKind.Tainted
                          && (a.Parameter.RefKind == RefKind.None
                              || a.Parameter.RefKind == RefKind.Ref
                              || a.Parameter.RefKind == RefKind.In));
             }
 
-            private void SetTaintedForEntity(IOperation operation, TaintedDataAbstractValue value)
+            private void CacheAbstractValueForBothOperationAndEntity(IOperation operation, TaintedDataAbstractValue value)
             {
+                CacheAbstractValue(operation, value);
                 if (AnalysisEntityFactory.TryCreate(operation, out AnalysisEntity analysisEntity))
                 {
-                    this.CurrentAnalysisData.SetAbstractValue(analysisEntity, value);
+                    // Cause we don't save every array as tainted, there's could be a new sanitized array entity we need to save.
+                    SetAbstractValueCore(CurrentAnalysisData, analysisEntity, value);
+                }
+            }
+
+            /// <summary>
+            /// Set array as tainted when there's <see cref="SourceInfo"/> taint all kinds of array before entering sink.
+            /// </summary>
+            /// <param name="target">This operation could be IArgumentOperation or IAssignmentOperation.Target.</param>
+            /// <param name="value">Operation that could produce array.</param>
+            private void UpdateAbstractValueForArrayBeforeEnteringSinkForArray(IOperation target, IOperation value)
+            {
+                if (value.Type is IArrayTypeSymbol arrayTypeSymbol)
+                {
+                    TaintedDataAbstractValue taintedDataAbstractValue = GetCachedAbstractValue(value);
+
+                    // Array in the default state: NotTainted.
+                    if (taintedDataAbstractValue.Kind == TaintedDataAbstractValueKind.NotTainted
+                        && this.DataFlowAnalysisContext.SourceInfos.IsSourceArray(arrayTypeSymbol, out TaintArrayKind taintArrayKind)
+                        && taintArrayKind == TaintArrayKind.All)
+                    {
+                        if (AnalysisEntityFactory.TryCreate(value, out AnalysisEntity analysisEntity))
+                        {
+                            // Array is not sanitized.
+                            if (!this.CurrentAnalysisData.TryGetValue(analysisEntity, out taintedDataAbstractValue))
```

Comment:
Move this comment inside the next if block, and mention we're relying on us not tracking AnalysisEntities unless they're sanitized or tainted.